### PR TITLE
[diffusion, ckpt] fix: restore checkpointed async rollout queue

### PR DESCRIPTION
**Pull request by Wx727**
### What does this PR do?

Port the checkpoint-boundary async recovery path from [verl#7037](https://github.com/verl-project/verl/pull/7037) to the independent Diffusion V1 trainer.

Before this change, Diffusion V1 restored the actor and dataloader state but not TransferQueue. In `separate_async` mode, this could lose queued prompt groups and trajectories after a restart.

This PR saves and restores TransferQueue, re-issues checkpointed in-flight prompts, preserves finished trajectories, and restores the configured warmup depth without overfilling it after recovery.

### Checklist Before Starting

- [x] Search for similar PRs: [verl-omni PR search](https://github.com/verl-project/verl-omni/pulls?q=is%3Apr+TransferQueue+checkpoint) and [verl#7037](https://github.com/verl-project/verl/pull/7037).
- [x] Format the PR title as `[{modules}] {type}: {description}`.
  - Proposed title: `[diffusion, ckpt] fix: restore checkpointed async rollout queue`

### Test

Tested with TransferQueue 0.1.9.

CPU regression tests:

```text
pytest -q tests/trainer/diffusion/test_v1_checkpoint_recovery_on_cpu.py
7 passed
```

Coverage includes:

- TransferQueue checkpoint capability detection by version and API availability;
- trainer save/load integration;
- in-flight recovery: re-issue `pending` and `running` groups, remove their partial trajectories, and leave finished groups untouched;
- a real TransferQueue checkpoint save/load/re-issue round trip;
- separate-async warmup top-up based on restored in-flight prompt groups.

Diffusion V1 separate-async GPU smoke:

```text
FlowGRPO diffusion v1 separate_async e2e test passed
```

Checkpoint recovery GPU smoke:

- saved actor, dataloader, and TransferQueue state at `global_step_1`;
- restarted with `trainer.resume_mode=auto`;
- restored checkpointed TransferQueue entries;
- continued training successfully;
- saved `global_step_2`.

### API and Usage Example

There is no user-facing trainer API or configuration change.

Existing resume configuration continues to be used:

```yaml
trainer:
  resume_mode: auto
  save_freq: 1
```

TransferQueue is upgraded from 0.1.8 to 0.1.9 because 0.1.9 is the first stable release providing the public `save_checkpoint` and `load_checkpoint` APIs.

### Design & Code Changes

- Save TransferQueue after the actor and dataloader state for non-sync trainers.
- Restore TransferQueue from the selected trainer checkpoint when available.
- Re-issue checkpointed `pending` and `running` prompt groups after entering the resumed training step.
- Clear partial trajectories before restarting an in-flight prompt group.
- Preserve finished prompt groups and trajectories for replay-buffer sampling.
- Top up separate-async warmup based on restored pending and running prompt groups.
- Guard checkpoint operations using both the TransferQueue version and API availability.
- Update project and CI TransferQueue pins from 0.1.8 to 0.1.9.

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl-omni/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/verl-project/verl-omni/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`.
- [x] Add / Update the documentation. Not applicable: this PR introduces no user-facing API or configuration change.
- [x] Add unit or end-to-end tests to cover the code. The new CPU regression suite is automatically collected by the existing CPU test workflow, and GPU smoke coverage was completed manually.

### CODE REVIEW

File: tests/trainer/diffusion/test_v1_checkpoint_recovery_on_cpu.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+
+def test_transfer_queue_checkpoint_guard(monkeypatch):
+    checkpoint_api = lambda *args, **kwargs: None
+    monkeypatch.setattr(trainer_base_module.tq, "save_checkpoint", checkpoint_api, raising=False)
+    monkeypatch.setattr(trainer_base_module.tq, "load_checkpoint", checkpoint_api, raising=False)
+
+    monkeypatch.setattr(trainer_base_module.tq, "__version__", "0.1.8", raising=False)
+    assert trainer_base_module._tq_supports_checkpoint() is False
+
+    monkeypatch.setattr(trainer_base_module.tq, "__version__", "0.1.9")
```
[the remark this answers is not in the document]
Comment (Wx727) [reply tests/trainer/diffusion/test_v1_checkpoint_recovery_on_cpu.py]: Updated to use raising=False consistently.

**Comment by zhtmike** [comment]
@knlnguyen1802

**Review by knlnguyen1802 — APPROVED** [review]
LGTM, nice catch up with upstream https://github.com/verl-project/verl/pull/7037

File: verl_omni/trainer/diffusion/v1/trainer_base.py [hunk 2]
```diff
@@ -1339,3 +1360,49 @@ def _load_checkpoint(self):
...
+            key
+            for key, tag in items.items()
+            if not tag.get("is_prompt", False) and self._trajectory_uid(key) in inflight_uid_set
+        ]
+        if partial_trajectory_keys:
+            tq.kv_clear(keys=partial_trajectory_keys, partition_id=partition_id)
+
+        tu.assign_non_tensor_data(batch, "global_steps", self.global_steps)
+        tags = [{"is_prompt": True, "status": "pending", "global_steps": self.global_steps} for _ in inflight_uids]
+        tq.kv_batch_put(keys=inflight_uids, partition_id=partition_id, tags=tags)
```
Comment (SamitHuang) [inline verl_omni/trainer/diffusion/v1/trainer_base.py]: `_submit_batch_to_rollout` always passes `fields=` for async mode so prompt tensors stay in TransferQueue. This tags-only `kv_batch_put` is a different write path: if 0.1.9 treats an omitted `fields` as a full-row replace, restored prompt data (images/latents/embeds) is wiped even though `generate_sequences` still has the in-memory batch from `kv_batch_get`.

Please either call `_submit_batch_to_rollout(batch)` here, or pass `fields=` the same way as the submit path, and assert in the real TQ roundtrip that prompt fields still exist after the tag update.

File: verl_omni/trainer/diffusion/v1/trainer_base.py [hunk 3]
```diff
@@ -1339,3 +1360,49 @@ def _load_checkpoint(self):
             self.train_dataloader.load_state_dict(torch.load(dataloader_path, weights_only=False))
         else:
             logger.warning(f"No dataloader state at {dataloader_path}, starting from scratch")
+
+        if self.trainer_mode != "sync" and _tq_supports_checkpoint():
```
Comment (SamitHuang) [inline verl_omni/trainer/diffusion/v1/trainer_base.py]: Missing `transfer_queue/` is silent, unlike the dataloader warning on the previous line. Pre-this-PR checkpoints and any resume where `_tq_supports_checkpoint()` is false will drop the async queue with no log, which is easy to misread as a successful recovery.

Please warn in both cases (absent snapshot, and TQ too old / APIs missing), matching the `data.pt` path.

File: verl_omni/trainer/diffusion/v1/trainer_separate_async.py [hunk 4]
```diff
@@ -231,6 +232,11 @@ def on_init_end(self):
         self.checkpoint_manager.update_weights(self.global_steps)
 
     def on_train_begin(self):
+        queue_state = tq.kv_list("train") or {}
+        if any(tag.get("is_prompt", False) for tag in queue_state.get("train", {}).values()):
```
Comment (SamitHuang) [inline verl_omni/trainer/diffusion/v1/trainer_separate_async.py]: This skips *all* warmup whenever any prompt row exists, including finished-only groups. After a correct dataloader restore, warmup would prefetch the *next* batches, not replay the same UIDs — so this is not duplicate suppression. Upstream [verl#7037](https://github.com/verl-project/verl/pull/7037) does not gate `on_train_begin` this way.

If restored `pending`/`running` groups already cover `num_warmup_batches`, skip is right. If the snapshot is only finished groups, or inflight count is smaller than warmup depth, this underfills the pipeline. Gate on inflight status (or `max(0, num_warmup_batches - restored_inflight_batches)`), not `is_prompt`.

**Review by SamitHuang — COMMENTED** [review]
Port of the verl#7037 async TQ resume path is the right fix, and the CPU suite covers capability detection plus inflight vs finished groups. Three issues before this should land:

1. Reissue updates tags with `kv_batch_put(..., tags=)` and skips `fields=`, unlike `_submit_batch_to_rollout`. Confirm 0.1.9 does not wipe prompt tensors; otherwise restored images/latents disappear after resume.
2. Missing `transfer_queue/` (old checkpoints) or `_tq_supports_checkpoint()==False` is silent — warn like `data.pt`.
3. `on_train_begin` skips all warmup if any prompt row exists, including finished-only. That is not UID-duplicate suppression; gate on inflight depth vs `num_warmup_batches`.

Also note: diffusion TQ snapshots can be large (latents/images). Worth a log of snapshot size on save.

**Comment by knlnguyen1802** [comment]
@Wx727 Please resolve other comment

File: verl_omni/trainer/diffusion/v1/trainer_base.py (hunk 2 above)
Comment (Wx727) [reply verl_omni/trainer/diffusion/v1/trainer_base.py]: Updated the reissue path to pass `fields=` using the same filtering as `_submit_batch_to_rollout`. The real TQ round-trip test now re-reads the prompt fields after reissue.

File: verl_omni/trainer/diffusion/v1/trainer_base.py (hunk 3 above)
Comment (Wx727) [reply verl_omni/trainer/diffusion/v1/trainer_base.py]: Added warnings for both unsupported TQ checkpoint APIs and a missing `transfer_queue/` snapshot. In both cases, the log states that the async queue starts empty.

File: verl_omni/trainer/diffusion/v1/trainer_separate_async.py (hunk 4 above)
Comment (Wx727) [reply verl_omni/trainer/diffusion/v1/trainer_separate_async.py]: Updated warmup to count only restored `pending`/`running` prompts and top up the configured depth. Added coverage for sufficient, finished-only, and partial in-flight states.

**Comment by Wx727** [comment]
@knlnguyen1802 Addressed and resolved the review comments in `ae5b9f6`. CPU regression tests: 7 passed; pre-commit checks on the changed files passed. Could you please re-add `ready-for-ci`? Thanks.

**Comment by SamitHuang** [comment]
Pls resolve the conflicts, thx

**Comment by Wx727** [comment]
@SamitHuang Conflicts have been resolved by rebasing onto the latest main. Could you please re-add the `ready-for-ci` label? Thanks!

**Comment by zhtmike** [comment]
Ping @knlnguyen1802 for this.

**Review by knlnguyen1802 — APPROVED** [review]
